Networking and gui

gadjevb opened the pull request:
All tasks are ready for review!

gadjevb commented:
I can't think of a way to test the server and the client.

freekman commented on src/main/java/com/clouway/task1/DownloadAgent.java:
```diff
@@ -0,0 +1,30 @@
+ */
+public class DownloadAgent {
+
+    public void downloadFile(String fileUrl, String targetUrl) throws Exception {
```
i dont see this part
``` така и начина по - който прогреса се обновява.```

gadjevb commented:
I don't know if this way of testing is Ok. I've created two classes, TestClient and TestServer. The logic behind them is the same(it does what it should), but I've added the fakeClient and fakeServer in them and changed the startServer and connect methods so that they can be testable!

freekman commented on src/main/java/com/clouway/task1/DownloadAgent.java:
```diff
@@ -0,0 +1,42 @@
+
+    private int counter;
+
+    public void downloadFile(String fileUrl, String targetUrl) throws Exception {
```
rename ```fileUrl``` to just ```url or src``` and the ```targetUrl```cab be renamed to dist

freekman replied:
still no progress bar to be found

freekman commented on src/main/java/com/clouway/task1/DownloadAgent.java:
```diff
@@ -0,0 +1,42 @@
+        OutputStream out = new BufferedOutputStream( new FileOutputStream(targetUrl + format));
+
+        int transfer = 0;
+        while((transfer = in.read()) != -1){
```
If i am not mistaking this way you are reading one bite at a time.

gadjevb commented:
If I'm not mistaken the GUI is no longer in the chapter.

freekman commented on src/main/java/com/clouway/task2/Server.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public void startServer(int port) throws IOException {
+        server = new ServerSocket(port);
```
where is the server socket closed

freekman commented on src/main/java/com/clouway/task2/Server.java:
```diff
@@ -0,0 +1,47 @@
+                        Date date = currentTimeAndDate.getTimeAndDate();
+                        out.println(date);
+                        connection.close();
+                        if(connection.isClosed()){
```
is this nesesery ,because you close the connectin on line 34
```js
 if(connection.isClosed()){
 +                            System.out.println("Client is offline!");
 +                        }else{
 +                            System.out.println("Client is online!");
 +                        }
```

gadjevb commented:
I understood that the server must stay up constantly, I will close it!

freekman commented on src/main/java/com/clouway/task2/CurrentTimeAndDate.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * @author Borislav Gadjev <gadjevb@gmail.com>
+ */
+public interface CurrentTimeAndDate {
```
Clock is a better  name

freekman commented on src/main/java/com/clouway/task2/Server.java:
```diff
@@ -0,0 +1,47 @@
+    private CurrentTimeAndDate currentTimeAndDate;
+    private ServerSocket server = null;
+    private Socket connection = null;
+    private PrintStream out = null;
```
do we need to close the ```PrintStream```

freekman commented on src/test/java/com/clouway/test/task2/ServerTest.java:
```diff
@@ -0,0 +1,64 @@
+ * @author Borislav Gadjev <gadjevb@gmail.com>
+ */
+public class ServerTest {
+    @Rule public JUnitRuleMockery context = new JUnitRuleMockery() {{setThreadingPolicy(new Synchroniser());}};
```
i think that you dont need to set setThreadingPolicy for this test because you are not using it .
use the format
```js
 @Rule 
  public JUnitRuleMockery .....
```

freekman commented on src/test/java/com/clouway/test/task2/ServerTest.java:
```diff
@@ -0,0 +1,64 @@
+ */
+public class ServerTest {
+    @Rule public JUnitRuleMockery context = new JUnitRuleMockery() {{setThreadingPolicy(new Synchroniser());}};
+    Clock clock = context.mock(Clock.class);
```
add  Access Modifier

freekman commented on src/test/java/com/clouway/test/task2/ServerTest.java:
```diff
@@ -0,0 +1,64 @@
+    Server server = new Server(clock);
+    String fromServer, result;
+
+    public void fakeClient(int port) throws IOException, InterruptedException {
```
it seams strange to have a method ```fakeClient()``` that actually connects to a server i think it would be better if you have a class with method connect
```js
class FakeClient{
....
private void connect(){
...
} 
}
```

freekman commented on src/test/java/com/clouway/test/task2/ServerTest.java:
```diff
@@ -0,0 +1,64 @@
+                    Socket client = new Socket("127.0.0.1", port);
+                    BufferedReader in = new BufferedReader(new InputStreamReader(client.getInputStream()));
+                    if((fromServer = in.readLine())!=null)
+                    {
```
.

gadjevb commented:
The only time I don't get an error is with setThreadingPolicy(new Synchroniser());

freekman commented on src/test/java/com/clouway/test/task2/ServerTest.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @Test
+    public void sendTimeAndDateToClient() throws IOException, InterruptedException, ParseException {
```
why not just happyPath

freekman commented on src/test/java/com/clouway/test/task2/ServerTest.java:
```diff
@@ -0,0 +1,64 @@
+
+        fakeClient(6000);
+        server.startServer(6000);
+        sleep(1000);
```
remove the sleep (to be sure that the server is started when the client is connecting you can use the Synchroniser and the display interface that we discussed )

freekman commented on src/main/java/com/clouway/task2/Client.java:
```diff
@@ -0,0 +1,33 @@
+public class Client {
+
+    public String connect(String host, int port) throws IOException, InterruptedException {
+        sleep(100);
```
.

freekman commented on src/main/java/com/clouway/task2/Client.java:
```diff
@@ -0,0 +1,33 @@
+ */
+public class Client {
+
+    public String connect(String host, int port) throws IOException, InterruptedException {
```
do you find it ok for a connect() method to return a string

freekman commented on src/main/java/com/clouway/task2/Server.java:
```diff
@@ -0,0 +1,43 @@
+    private Clock clock;
+    private ServerSocket server = null;
+    private Socket connection = null;
+    private PrintStream out = null;
```
is there a reason these three not to be declared in the start method

freekman commented on src/main/java/com/clouway/task2/Server.java:
```diff
@@ -0,0 +1,43 @@
+        this.screen = screen;
+    }
+
+    public void startServer(int port) throws IOException {
```
just ```start(...)``` in enough

freekman commented on src/test/java/com/clouway/test/task2/ServerTest.java:
```diff
@@ -0,0 +1,87 @@
+                        if((fromServer = in.readLine())!=null)
+                        {
+                            message = fromServer;
+                            screen.display("Time and date received!");
```
necessary

freekman commented on src/test/java/com/clouway/test/task2/ServerTest.java:
```diff
@@ -0,0 +1,87 @@
+                        client = new Socket(host, port);
+                        BufferedReader in = new BufferedReader(new InputStreamReader(client.getInputStream()));
+                        if((fromServer = in.readLine())!=null)
+                        {
```
format code

gadjevb commented:
Yes, otherwise 1/10 test fails, because the getMessage() methods runs before the actual message was received.

freekman commented on src/main/java/com/clouway/task2/Client.java:
```diff
@@ -0,0 +1,44 @@
+        this.screen = screen;
+    }
+
+    public synchronized void connect(String host,int port) throws InterruptedException {
```
why is this method synchronized

freekman commented on src/main/java/com/clouway/task2/Client.java:
```diff
@@ -0,0 +1,44 @@
+                    client = new Socket(host, port);
+                    BufferedReader in = new BufferedReader(new InputStreamReader(client.getInputStream()));
+                    if((fromServer = in.readLine())!=null)
+                    {
```
format your code everywhere

freekman commented:
Please remove all the commented code and for the task 3 remove the not needed logic

freekman commented on src/main/java/com/clouway/task3/Screen.java:
```diff
@@ -0,0 +1,8 @@
+ * @author Borislav Gadjev <gadjevb@gmail.com>
+ */
+public interface Screen {
+    void display(String message) throws NoSocketException;
```
why does the screen throw a NoSocketException

gadjevb commented:
Because the screen displays the message, and that way I can check for the exception. Otherwise I must handle it in the thread.

freekman commented:
It is a bad practice for a interface method to throw exceptions
this way you are showing that the display is working with sockets if tomorrow  you decide not to use sockets then what?

gadjevb commented:
I will fix it, I will use screen.display("Server is offline!"); to check for it in the catch(NoSocketException ){}

freekman commented on src/test/java/com/clouway/test/task3/ServerWithMultipleClientsTest.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    @Test
+    public void happyPathWithOneClient() throws IOException, InterruptedException, NoSocketException {
```
:) no please just happyPath

freekman commented on src/test/java/com/clouway/test/task3/ServerWithMultipleClientsTest.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    @Test
+    public void happyPathWithSeveralClients() throws IOException, InterruptedException, NoSocketException {
```
multipleConnections will be ok

freekman commented on src/test/java/com/clouway/test/task3/ServerWithMultipleClientsTest.java:
```diff
@@ -0,0 +1,94 @@
+
+    @Test
+    public void happyPathWithOneClient() throws IOException, InterruptedException, NoSocketException {
+        final States connecting = context.states("Waiting for connection!");
```
rename to state

freekman commented on src/main/java/com/clouway/task3/Server.java:
```diff
@@ -0,0 +1,71 @@
+      }
+      try {
+          sendGreeting(clients.get((clients.size() - 1)), "Hello, you're client №" + clients.size() + " in the list!");
+          screen.display("Hello, you're client №" + clients.size() + " in the list!");
```
extract ```"Hello, you're client №" + clients.size() + " in the list!"``` in a variable to reduce duplication

freekman commented on src/main/java/com/clouway/task3/Server.java:
```diff
@@ -0,0 +1,71 @@
+    private void sendInformation() {
+      for (int i = 0; i < (clients.size() - 1); i++) {
+          try {
+              sendGreeting(clients.get(i), "There's a new client in the list with №" + (clients.size()) + "!");
```
Where is this tested

freekman commented on src/main/java/com/clouway/task3/Client.java:
```diff
@@ -0,0 +1,41 @@
+ * @author Borislav Gadjev <gadjevb@gmail.com>
+ */
+public class Client {
+    private BufferedReader in = null;
```
Why is it declared here and not in thee connect method

freekman commented on src/main/java/com/clouway/task3/Client.java:
```diff
@@ -0,0 +1,41 @@
+        }.start();
+    }
+
+    private void hasReceivedMessage() throws IOException, NoSocketException {
```
very bad method name what is an void ```hasReceivedMessage() ```

freekman commented on src/main/java/com/clouway/task3/Client.java:
```diff
@@ -0,0 +1,41 @@
+    private BufferedReader in = null;
+    private Screen screen;
+
+    public Client(Screen screen) {
```
i cant see this part of the task
```js
Стартира безкраен цикъл за четене на съобщения.
```

freekman commented on src/test/java/com/clouway/test/task3/ClientTest.java:
```diff
@@ -0,0 +1,121 @@
+
+    @Test
+    public void serverGoesOffline() throws IOException, InterruptedException {
+        FakeServer badServer = new FakeServer();
```
just server

gadjevb commented:
BufferedReader and the endless cycle are in the receiveMessage() method, that's why BufferedReader is not declared in the connect() method!

freekman commented:
are you shure

gadjevb commented:
Ok, now the cycle is truly endless.

freekman commented:
LGTM
